src/mipmap.rs

```diff
@@ -0,0 +1,297 @@
+///
+/// These are created from a `PixmapRef` as a base source which can be fetched
+/// using level `0`
+///
+#[derive(Debug)]
+pub struct Mipmaps<'a> {
+    levels: Vec<Pixmap>,
+    base_pixmap: PixmapRef<'a>,
+}
+
+impl<'a> Mipmaps<'a> {
+    /// Allocates a new set of mipmaps from a base pixmap
+    pub fn new(p: PixmapRef<'a>) -> Self {
+        Mipmaps {
+            levels: Vec::new(),
+            base_pixmap: p,
+        }
+    }
+
+    /// Fetch a mipmap to be used - or base pixmap if zero is given
+    pub fn get(&self, level: usize) -> PixmapRef {
+        return if level > 0 {
+            self.levels.get(level - 1).unwrap().as_ref()
+        } else {
+            self.base_pixmap
+        };
+    }
+
+    /// Ensure this many levels of mipmap are available, returning
+    /// an index to be used with get()
+    pub fn build(&mut self, required_levels: usize) -> usize {
+        let mut src_level = self.levels.len();
+        let mut src_pixmap = self.get(src_level);
+        let mut level_width = src_pixmap.width();
+        let mut level_height = src_pixmap.height();
+
+        while src_level < required_levels {
+            level_width = (level_width as f32 / 2.0).floor() as u32;
+            level_height = (level_height as f32 / 2.0).floor() as u32;
+
+            // Scale image down
+            let mut dst_pixmap = Pixmap::new(level_width, level_height).unwrap();
+            let dst_width = dst_pixmap.width() as usize;
+            let dst_height = dst_pixmap.height() as usize;
+            let dst_pixels = dst_pixmap.pixels_mut();
+
+            let src_pixels = src_pixmap.pixels();
+            let src_width = src_pixmap.width() as usize;
+            let src_height = src_pixmap.height() as usize;
+
+            //  To produce each mip level, we need to filter down by 1/2 (e.g. 100x100 -> 50,50)
+            //  If the starting dimension is odd, we floor the size of the lower level (e.g. 101 -> 50)
+            //  In those (odd) cases, we use a triangle filter, with 1-pixel overlap between samplings,
+            //  else for even cases, we just use a 2x box filter.
+            //
+            //  This produces 4 possible isotropic filters: 2x2 2x3 3x2 3x3 where WxH indicates the number of
+            //  src pixels we need to sample in each dimension to produce 1 dst pixel.
+            let mut downsample: fn(
+                &[PremultipliedColorU8],
+                usize,
+                usize,
+                &mut [PremultipliedColorU8],
+                usize,
+                usize,
+            ) = downsample_2_2;
+
+            if src_height & 1 == 1 {
+                if src_width & 1 == 1 {
+                    downsample = downsample_3_3;
+                } else {
+                    downsample = downsample_2_3;
+                }
+            } else {
+                if src_width & 1 == 1 {
+                    downsample = downsample_3_2;
+                }
+            }
```

Comment:
I think this whole block and the `let mut downsample` with a type annotation can be reduced to something like:

```
    let downsample = match (src_width & 1 == 0, src_height & 1 == 0) {
        (true, true) => downsample_2_2,
        (true, false) => downsample_2_3,
        (false, true) => downsample_3_2,
        (false, false) => downsample_3_3,
    };
```

Reply:
Certainly much cleaner!